runtime/src/main/java/org/corfudb/runtime/collections/Query.java

```diff
@@ -0,0 +1,424 @@
+                .map(v -> Optional.ofNullable(joinProjection)
+                        .map(function -> function.apply(v))
+                        .orElse((U) v))
+                .collect(Collectors.toList()));
+    }
+
+    /**
+     * Merge Function which combines the result two tables at a time.
+     *
+     * @param <R> Type of result.
+     */
+    @FunctionalInterface
+    public interface MergeFunction<R> {
+
+        /**
+         * Merges the specified argument list.
+         *
+         * @param arguments Objects across tables to be merged.
+         * @return Resultant object.
+         */
+        R merge(List<Object> arguments);
+    }
+
+    /**
+     * Combines the result of a multi table join, two tables at a time
+     * @param list
+     * @param mergePayload
+     * @param func
+     * @param depth
+     * @param <R>
+     * @return
+     */
+    private <R> List<R> merge(@Nonnull List<Collection<?>> list,
+                              @Nonnull List<Object> mergePayload,
+                              @Nonnull MergeFunction<R> func,
+                              int depth) {
+        Collection<?> collection = list.get(depth);
+        if (list.size() - 1 == depth) {
+            List<R> result = new ArrayList<>();
+            for (Object o : collection) {
+                List<Object> finalMergeList = new ArrayList<>(mergePayload);
+                finalMergeList.add(o);
+                R mergeResult = func.merge(finalMergeList);
+                if (mergeResult != null) {
+                    result.add(mergeResult);
+                }
+            }
+            return result;
+        }
+
+        List<R> result = new ArrayList<>();
+        for (Object o : collection) {
+            List<Object> mergeList = new ArrayList<>(mergePayload);
+            mergeList.add(o);
+            result.addAll(merge(list, mergeList, func, depth + 1));
+        }
+        return result;
+    }
+
+    /**
+     * Performs join of multiple tables.
+     *
+     * @param tableNames   Collection of table names to be joined.
+     * @param joinFunction MergeFunction to perform the join across the specified tables.
+     * @param <R>          Type of resultant Object.
+     * @return Result of the query.
+     */
+    @Nonnull
+    public <R> QueryResult<R> executeMultiJoinQuery(@Nonnull final Collection<String> tableNames,
+                                                    @Nonnull final MergeFunction<R> joinFunction) {
+
+        List<Collection<? extends Object>> values = new ArrayList<>();
+        for (String tableName : tableNames) {
+            Collection<Message> messages = scanAndFilter(tableName, null, corfuRecord -> true)
+                    .stream()
+                    .map(corfuRecord -> corfuRecord.getPayload())
+                    .collect(Collectors.toList());
+            values.add(messages);
+        }
+
+        Collection<R> mergedResults = merge(values, new ArrayList<>(), joinFunction, 0);
```

Comment:
what does 0 mean?

Reply:
The last parameter is the depth of the multi table join. In the first cut we are only going to support merging 2 tables at a time, perhaps in the future we can support multi table joins, but for now, we are just passing in 0.